core/src/machine/ppu.rs

```diff
@@ -356,69 +593,107 @@ impl Ppu {
+        self.fetch_map_line_offset = map_offset + 32 * tile_y as u16;
     }
 
     /// Performs one step of the pixel transfer phase. This involves fetching
     /// new tile data and emitting the pixels.
-    fn pixel_transfer_step(&mut self, display: &mut impl Display) {
+    fn pixel_transfer_step(
+        &mut self,
+        display: &mut impl Display,
+        interrupt_controller: &mut InterruptController,
+    ) {
         // Push out up to four new pixels if we have enough data in the FIFO.
         let mut pixel_pushed = 0;
         while self.fifo.len() > 8 && pixel_pushed < 4 {
-            self.push_pixel(display);
+            // Check if the next pixel should be discarded or not.
+            if self.num_throw_away_pixels == 0 {
+                self.push_pixel(display);
+
+                // We just bumped our `current_column`. Now we check if we
+                // reached the start of the window.
+                let window_trigger = self.regs().is_window_enabled()
+                    && self.regs().scroll_win_x.get() >= 7
+                    && self.current_column == self.regs().scroll_win_x.get() - 7;
+                if window_trigger {
+                    // We need to basically reset the whole state.
+                    self.fifo.clear();
+                    self.fetch_started = false;
+
+                    // The following is nearly the same as in
+                    // `prepare_pixel_transfer`.
+                    self.fetch_tile_x = 0;
+
+                    // Calculate the Y position and the offset within one tile.
+                    let pos_y = (self.regs().scroll_win_y + self.regs().current_line).get();
+                    self.line_in_tile_offset = (pos_y % 8) * 2;
+
+                    // Here we precompute the address offset to the first tile
+                    // of the current line in the tile map. This means that we
+                    // can easily compute the address of the real tile later as
+                    // `fetch_map_line_offset + fetch_tile_x`.
+                    let tile_y = pos_y / 8;
+                    let map_offset = self.regs().window_tile_map_address().start();
+                    self.fetch_map_line_offset = map_offset + 32 * tile_y as u16;
+                }
+            } else {
+                let _ = self.fifo.emit();
+                self.num_throw_away_pixels -= 1;
+            }
+
             pixel_pushed += 1;
 
             // We are at the end of the line, stop everything and go to
             // H-Blank.
             if self.current_column == SCREEN_WIDTH as u8 {
-                self.set_phase(Phase::HBlank);
+                self.registers.set_mode(Mode::HBlank);
+
+                // Trigger H-Blank interrupt if enabled.
+                if self.regs().hblank_interrupt() {
+                    interrupt_controller.request_interrupt(Interrupt::LcdStat);
+                    debug!("triggered");
```

Comment:
This is a bit meaningless. Perhaps remove the output (or enhance it)? For consistency: The other interrupts do not trigger an output.